mdsavian/SAVIAN_fe_MARLON_exercise
# improvements

# Teams.tsx

-   we are using typescript, so we need to type everything on this page, that way we can get type errors earlier, like the `isLoading` prop on the List component

    function returns
    state
    the component `React. FC`

-   change var to const on the `MapT` function, since this is a function and will not change, we must use const.

-   change the var `columns` to be a const -- we can remove the var and assign it directly to the returned object, but I prefer to leave it this way, it's more descriptive

-   change the name of the function `MapT` to something more descriptive, like `MapTeamsList`

-   Change the getTeams to treat error case
    define an error state, show the error on the page, otherwise the loading spinner will be showing forever

## Teams. test

-   change the name of the test from `testTeams` to `teams.test`

-   remove the fake timers, we don't need them on the test, we are not showing any date or time

-   instead of using a comment with the TODO, I prefer to use `it.todo`, so the test will not run

-   remove unused imports (fireEvent, act)

-   implement the render error test

-   implement the render spinner test

-   test if the backButton is not displayed

-   test if the title is displayed

# TeamOverview

-   change var `mapArray` and `mapTLead` to const
-   rename `mapArray` to `mapUsers`
-   rename `mapTLeam` to `mapTeamLead`
-   type the `tLead` var
-   type function returns
-   used a more descriptive name on user.map
-   used Promise.All to fetch the member data so we can call the get in parallel

## TeamOverview.test

-   rename to `teamOverview.test`
-   remove fake timers
-   fix useLocation state
-   create a test to check the render team lead
-   improve test to check render users and team lead
-   check loading spinner
-   check the header with the back button and team name

# UserOverview

-   remove `mapU` function
-   created `mapUserColumns` to reduce the duplicated code
-   rename test to `userOverview.test`

# List.tsx

-   change isLoading prop to boolean
-   change the test name to `list.test`

# Card.tsx

-   change columns prop to be a ListItemColumn Array
-   extract the onClick to a function, so the code is more organized
-   change the test name to `card.test`

# Header.tsx

-   removed the styled `Title` component because we don't have any styles. Replaced with `<h1>` tag
-   remove the backArrow icon and use an SVG from material ui
-   remove borders, and background color from the back button

## Header.test

-   change the name to `header.test`
-   remove unused fake timers

# Creating a Search field

-   Created a test to render the search field
-   Created the style folder to create the styled input.
-   Created the SearchField component and use the styled input
-   Implement a test to check if the setSearchTerm was called on onChange
-   Received the setSearchTerm as a parameter and called it on the onChange

-   Wrote a test to filter teams by name
-   Used the SearchField on the Teams component
-   Created the search term state on the teams component
-   Filtered the teams based on the search term

-   Wrote a test to filter users by name
-   Used the SearchField on the TeamOverview component
-   Created the search term state on the teamOverview component
-   Filtered the users based on the search term

-   Wrote a test to show not found message when items are empty on the List component
-   Implement the not found message on the List component
-   Wrote a test to show users not found on the TeamOverview component
-   Pass an empty message parameter to the List component on the TeamOverview component
-   Wrote a test to show users not found on the Teams component
-   Pass an empty message parameter to the List component on the Teams component
